README
A port of Ruby for OpenVMS based on Ruby 1.8.2 (MRI).

This work is early Alpha and not suitable for release yet. See TODO
for more details.


ruby/changelog
Fri Dec 17 13:50:00 2004  Akiyoshi, Masamichi (akiyoshi@hp.com)

	* vms/vmsruby_private.c, vms/vmsruby_private.h: private routines 
	  for VMS port are added.

	* eval.c (ruby_init): change to call VMS private intialization routine.

Fri Dec 17 13:33:58 2004  Nobuyoshi Nakada  <nobu@ruby-lang.org>

	* lib/cgi/session.rb (CGI::Session#initialize): control adding
	  session_id hidden fields.  fixed: [ruby-talk:123850]

Thu Dec 16 23:25:25 2004  Masatoshi SEKI  <m_seki@mva.biglobe.ne.jp>

	* lib/drb/drb.rb, lib/drb/ssl.rb: backported from CVS HEAD.
	  [druby-ja:101]

	* test/drb/test_drb.rb: adjust and reduce sleep (backported from
	  CVS HEAD.)

Thu Dec 16 18:44:58 2004  GOTOU Yuuzou  <gotoyuzo@notwork.org>

	* lib/webrick/httpserver.rb (WEBrick::HTTPServer#run): should wait
	  for reading request till data arrive. [ruby-talk:121068]

	* lib/webrick/server.rb (WEBrick::GenericServer#start_thread):
	  should log about all accepted socket. [ruby-core:03962]

	* lib/webrick/accesslog.rb (WEBrick::AccessLog#setup_params):
	  "%%" and "%u" are supported. [webricken:135]

	* lib/webrick/httpservlet/filehandler.rb
	  (WEBrick::HTTPServlet::FileHandler#check_filename):
	  :NondisclosureName is acceptable if it is Enumerable.

	* lib/webrick/config.rb (WEBrick::Config::FileHandler):
	  default value of :NondisclosureName is [".ht*", "*~"].

Thu Dec 16 18:36:52 2004  GOTOU Yuuzou  <gotoyuzo@notwork.org>

	* ext/openssl/ossl.c (ossl_raise): refine message format.

Thu Dec 16 16:29:44 2004  Hidetoshi NAGAI  <nagai@ai.kyutech.ac.jp>

	* ext/tk/sample/demos-en/widget: modify version check for
	  supporting features

Thu Dec 16 16:03:50 2004  Hidetoshi NAGAI  <nagai@ai.kyutech.ac.jp>

	* ext/tk/lib/tk/bindtag.rb: bug fix [ruby-talk: 123667]

	* ext/tk/lib/tk/timer.rb: accept :idle for the interval argument

	* ext/tk/lib/tk.rb: add TkComm._callback_entry?()

	* ext/tk/lib/multi-tk.rb: add MultiTkIp.cb_entry_class

	* ext/tk/lib/tk/canvas.rb: use TkComm._callback_entry?()

	* ext/tk/lib/tk/canvastag.rb: ditto

	* ext/tk/lib/tk/dialog.rb: ditto

	* ext/tk/lib/tk/optiondb.rb: ditto

	* ext/tk/lib/tk/text.rb: ditto

	* ext/tk/lib/tk/texttag.rb: ditto

	* ext/tk/lib/tk/textwindow.rb: ditto

	* ext/tk/lib/tk/timer.rb: ditto

	* ext/tk/lib/tk/validation.rb: ditto

	* ext/tk/lib/tkextlib/*: ditto

Thu Dec 16 03:14:28 2004  Minero Aoki  <aamine@loveruby.net>

	* lib/net/http.rb (basic_encode): return value of pack('m') may
	  include multiple CR/LFs.  Backported from main trunk (rev 1.112).
	  [ruby-dev:25212]

Wed Dec 15 18:48:42 2004  Shugo Maeda  <shugo@ruby-lang.org>

	* ext/curses/curses.c (window_subwin): call NUM2INT() before
	GetWINDOW(). (backported from CVS HEAD)

Wed Dec 15 17:03:50 2004  NAKAMURA Usaku  <usa@ruby-lang.org>

	* win32/win32.[ch] (rb_w32_isatty): new function to replace MSVCRT's
	  isatty because it never sets errno. (backported from CVS HEAD)

Wed Dec 15 15:39:32 2004  GOTOU Yuuzou  <gotoyuzo@notwork.org>

	* ext/openssl/ossl_x509name.c (ossl_x509name_to_a): avoid SEGV
	  (rollback the previous commit).

Wed Dec 15 14:22:10 2004  NAKAMURA Usaku  <usa@ruby-lang.org>

	* win32/win32.c (rb_w32_fdisset): check whether the handle is valid.
	  fixed: [ruby-core:03959]

Wed Dec 15 10:30:37 2004  Yukihiro Matsumoto  <matz@ruby-lang.org>

	* ext/openssl/ossl_digest.c (ossl_digest_initialize): [ruby-dev:25198]

Tue Dec 14 17:10:09 2004  NAKAMURA Usaku  <usa@ruby-lang.org>

	* win32/win32.c (rb_w32_close): need to reset osfhnd().

Tue Dec 14 14:03:57 2004  GOTOU Yuuzou  <gotoyuzo@notwork.org>

	* ext/openssl/ossl.c (ossl_raise): avoid buffer overrun.
	  [ruby-dev:25187]

Tue Dec 14 12:36:04 2004  Yukihiro Matsumoto  <matz@ruby-lang.org>

	* lib/cgi/session.rb (CGI::Session::initialize): generate new
	  session if given session_id does not exist.  [ruby-list:40368]

Mon Dec 13 18:13:52 2004  Tanaka Akira  <akr@m17n.org>

	* gc.c (stack_end_address): new function to obtain stack end address.
	  stack_end_address calls __builtin_frame_address(0) to obtain the
	  frame pointer of a stack frame of stack_end_address.  The address
	  is the stack pointer of the caller's stack frame.
	  (SET_STACK_END): use stack_end_address.
	  This makes the conservative garbage collector to scan a stack frame
	  of the garbage_collect function itself.  This is required because
	  callee-save registers may be stored in the frame.
	  [ruby-dev:25158]

Mon Dec 13 00:58:02 2004  Tanaka Akira  <akr@m17n.org>

	* lib/pathname.rb (cleanpath_aggressive): make it private.
	  (cleanpath_conservative): ditto.
	  Suggested by Daniel Berger.  [ruby-core:3914]

Sun Dec 12 20:06:38 2004  Masatoshi SEKI  <m_seki@mva.biglobe.ne.jp>

	* lib/drb/drb.rb: backported from CVS HEAD.

Sun Dec 12 10:35:10 2004  Dave Thomas  <dave@pragprog.com>

	* lib/rdoc/generators/template/html/html.rb (RDoc::Page): Don't
	  show an accessor's r/w flag if none was specified

Sun Dec 12 10:14:03 2004  Dave Thomas  <dave@pragprog.com>

	* lib/rdoc/rdoc.rb (RDoc::RDoc::parse_files): Never exclude files
	  explicitly given on the command line.

Sat Dec 11 20:12:21 2004  Masatoshi SEKI  <m_seki@mva.biglobe.ne.jp>

	* lib/drb/drb.rb: add DRbRemoteError. [ruby-list:40348],
	  [ruby-list:40390]

	* test/drb/drbtest.rb: ditto.

	* test/drb/ut_drb.rb: ditto.

Sat Dec 11 15:38:14 2004  Yukihiro Matsumoto  <matz@ruby-lang.org>

	* lib/jcode.rb (String::succ): [ruby-dev:25156]

Sat Dec 11 12:41:55 2004  NAKAMURA Usaku  <usa@ruby-lang.org>

	* eval.c (run_trap_eval): prototype; avoid VC++ warnings.

	* ext/socket/getaddrinfo.c: fix typo. fixed: [ruby-core:03947]

	* win32/win32.c: need to include dln.h.

Sat Dec 11 00:10:18 2004  Yukihiro Matsumoto  <matz@ruby-lang.org>

	* io.c (io_reopen): [ruby-dev:25150]

Fri Dec 10 08:39:27 2004  Nobuyoshi Nakada  <nobu@ruby-lang.org>


ruby/lex.c
/* C code produced by gperf version 2.7.2 */
/* Command-line: gperf -p -j1 -i 1 -g -o -t -N rb_reserved_word -k'1,3,$' ./keywords  */
struct kwtable {char *name; int id[2]; enum lex_state state;};

#define TOTAL_KEYWORDS 40
#define MIN_WORD_LENGTH 2
#define MAX_WORD_LENGTH 8
#define MIN_HASH_VALUE 6
#define MAX_HASH_VALUE 55
/* maximum key range = 50, duplicates = 0 */

#ifdef __GNUC__
__inline
#else
#ifdef __cplusplus
inline
#endif
#endif
static unsigned int
hash (str, len)
     register const char *str;
     register unsigned int len;
{
  static unsigned char asso_values[] =
    {
      56, 56, 56, 56, 56, 56, 56, 56, 56, 56,
      56, 56, 56, 56, 56, 56, 56, 56, 56, 56,
      56, 56, 56, 56, 56, 56, 56, 56, 56, 56,
      56, 56, 56, 56, 56, 56, 56, 56, 56, 56,
      56, 56, 56, 56, 56, 56, 56, 56, 56, 56,
      56, 56, 56, 56, 56, 56, 56, 56, 56, 56,
      56, 56, 56, 11, 56, 56, 36, 56,  1, 37,
      31,  1, 56, 56, 56, 56, 29, 56,  1, 56,
      56, 56, 56, 56, 56, 56, 56, 56, 56, 56,
      56, 56, 56, 56, 56,  1, 56, 32,  1,  2,
       1,  1,  4, 23, 56, 17, 56, 20,  9,  2,
       9, 26, 14, 56,  5,  1,  1, 16, 56, 21,
      20,  9, 56, 56, 56, 56, 56, 56, 56, 56,
      56, 56, 56, 56, 56, 56, 56, 56, 56, 56,
      56, 56, 56, 56, 56, 56, 56, 56, 56, 56,
      56, 56, 56, 56, 56, 56, 56, 56, 56, 56,
      56, 56, 56, 56, 56, 56, 56, 56, 56, 56,
      56, 56, 56, 56, 56, 56, 56, 56, 56, 56,
      56, 56, 56, 56, 56, 56, 56, 56, 56, 56,
      56, 56, 56, 56, 56, 56, 56, 56, 56, 56,
      56, 56, 56, 56, 56, 56, 56, 56, 56, 56,
      56, 56, 56, 56, 56, 56, 56, 56, 56, 56,
      56, 56, 56, 56, 56, 56, 56, 56, 56, 56,
      56, 56, 56, 56, 56, 56, 56, 56, 56, 56,
      56, 56, 56, 56, 56, 56, 56, 56, 56, 56,
      56, 56, 56, 56, 56, 56
    };
  register int hval = len;

  switch (hval)
    {
      default:
      case 3:
        hval += asso_values[(unsigned char)str[2]];
      case 2:
      case 1:
        hval += asso_values[(unsigned char)str[0]];
        break;
    }
  return hval + asso_values[(unsigned char)str[len - 1]];
}

#ifdef __GNUC__
__inline
#endif
struct kwtable *
rb_reserved_word (str, len)
     register const char *str;
     register unsigned int len;
{
  static struct kwtable wordlist[] =
    {
      {""}, {""}, {""}, {""}, {""}, {""},
      {"end", {kEND, kEND}, EXPR_END},
      {"else", {kELSE, kELSE}, EXPR_BEG},
      {"case", {kCASE, kCASE}, EXPR_BEG},
      {"ensure", {kENSURE, kENSURE}, EXPR_BEG},
      {"module", {kMODULE, kMODULE}, EXPR_BEG},
      {"elsif", {kELSIF, kELSIF}, EXPR_BEG},
      {"def", {kDEF, kDEF}, EXPR_FNAME},
      {"rescue", {kRESCUE, kRESCUE_MOD}, EXPR_MID},
      {"not", {kNOT, kNOT}, EXPR_BEG},
      {"then", {kTHEN, kTHEN}, EXPR_BEG},
      {"yield", {kYIELD, kYIELD}, EXPR_ARG},
      {"for", {kFOR, kFOR}, EXPR_BEG},
      {"self", {kSELF, kSELF}, EXPR_END},
      {"false", {kFALSE, kFALSE}, EXPR_END},
      {"retry", {kRETRY, kRETRY}, EXPR_END},
      {"return", {kRETURN, kRETURN}, EXPR_MID},
      {"true", {kTRUE, kTRUE}, EXPR_END},
      {"if", {kIF, kIF_MOD}, EXPR_BEG},
      {"defined?", {kDEFINED, kDEFINED}, EXPR_ARG},
      {"super", {kSUPER, kSUPER}, EXPR_ARG},
      {"undef", {kUNDEF, kUNDEF}, EXPR_FNAME},
      {"break", {kBREAK, kBREAK}, EXPR_MID},
      {"in", {kIN, kIN}, EXPR_BEG},
      {"do", {kDO, kDO}, EXPR_BEG},
      {"nil", {kNIL, kNIL}, EXPR_END},
      {"until", {kUNTIL, kUNTIL_MOD}, EXPR_BEG},
      {"unless", {kUNLESS, kUNLESS_MOD}, EXPR_BEG},
      {"or", {kOR, kOR}, EXPR_BEG},
      {"next", {kNEXT, kNEXT}, EXPR_MID},
      {"when", {kWHEN, kWHEN}, EXPR_BEG},
      {"redo", {kREDO, kREDO}, EXPR_END},
      {"and", {kAND, kAND}, EXPR_BEG},
      {"begin", {kBEGIN, kBEGIN}, EXPR_BEG},
      {"__LINE__", {k__LINE__, k__LINE__}, EXPR_END},
      {"class", {kCLASS, kCLASS}, EXPR_CLASS},
      {"__FILE__", {k__FILE__, k__FILE__}, EXPR_END},
      {"END", {klEND, klEND}, EXPR_END},
      {"BEGIN", {klBEGIN, klBEGIN}, EXPR_END},
      {"while", {kWHILE, kWHILE_MOD}, EXPR_BEG},
      {""}, {""}, {""}, {""}, {""}, {""}, {""}, {""}, {""},
      {""},
      {"alias", {kALIAS, kALIAS}, EXPR_FNAME}
    };

  if (len <= MAX_WORD_LENGTH && len >= MIN_WORD_LENGTH)
    {
      register int key = hash (str, len);

      if (key <= MAX_HASH_VALUE && key >= 0)
        {
          register const char *s = wordlist[key].name;

          if (*str == *s && !strcmp (str + 1, s + 1))
            return &wordlist[key];
        }
    }
  return 0;
}
